Properly format times in automatically generated subjects

NobodysNightmare opened the pull request:
Using default formatting rules for times, so depending on the instance setting, they will be formatted as "18h" or "2d 2h".

# Ticket
https://community.openproject.org/wp/63477

NobodysNightmare committed:
Properly format times in automatically generated subjects

Using default formatting rules for times, so depending on the instance
setting, they will be formatted as "18h" or "2d 2h".

Kharonus commented on app/models/work_package_types/pattern_resolver.rb:
```diff
@@ -57,32 +57,34 @@ def get_value(work_package, pattern_token)
+        return ATTRIBUTE_PLACEHOLDER
       end
+
+      return DurationConverter.output(value) if key.to_s.ends_with?("_time")
```
🟡 I am very unhappy with that solution right now, but also can't think yet of a better one.

My problem is we add here some "magic", which is completely unseen on the "other side", where those keys are defined. What if we have now a new attribute called `maintenance_period` or `upkeep_duration`? we also want to render this with the duration converter, right? What if we have something that is stored as days or years, instead of hours?

I hate this :/ For now, this fix is functional, only because we have two keys ending with `_time` and both are describing hours.

So, shall we wait until this is broken? Or should we try to introduce now something more elaborate?

Kharonus replied:
One possibility would be to not have the "stringify" responsibility tied to the resolver, but actually make it part of the token callback itself. Right now we actually never need the "value" of an attribute, but always the string representation of it.

Kharonus replied:
Opinions?

NobodysNightmare replied:
Let me dissect your concerns:

> I am very unhappy with that solution right now

Agreed. That's understandable. I was looking for a straight forward fix, but I am open for a deeper "root cause" fix".

> One possibility would be to not have the "stringify" responsibility tied to the resolver, but actually make it part of the token callback itself

I like this approach. I didn't look too deep into the specifics of what this will mean in the end, but I can imagine that this could work. In general I think it's helpful to have "type" information in one way or another when trying to get a string representation. Right now we don't quite have that (which led to the ugly fix).

> What if we have something that is stored as days or years, instead of hours?

I wouldn't go as far as that right now to be honest. Current usage of this existing helper class is evidence enough for me that we so far only store durations as number of hours. If we want to store a duration as something else or if we want to enable pattern formatting of something that doesn't play by the rules we defined today, we'll have to revisit this. But right now I don't see any need to act here, because we play by the rules (i.e. interfaces) laid out beforehand. In other words: Our implementation here doesn't add new problems to this.

NobodysNightmare replied:
> One possibility would be to not have the "stringify" responsibility tied to the resolver, but actually make it part of the token callback itself

Adendum: I am currently trying to figure out how this proposal could look in practice. What's interesting is that **we partially already do this**. Looking at values like `assignee`, `author`, etc, we do not return the `User` and leave the formatting up to `stringify`. Instead we decide for the formatting in the property mapper and return the name only.

NobodysNightmare replied:
More findings (making good progress towards a solution):

```
      def context
        case key.to_s
        when /^project_/
          :project
        when /^parent_/
          :parent
        else
          :work_package
        end
      end
```

Look! Code that magically uses the `key` to make decisions...

NobodysNightmare replied:
@Kharonus Refactoring is completed. Tell me what you think.

NobodysNightmare committed:
Pull formatting into attribute token and mapper

When addressing formatting concerns for durations in the
previous commit, we realized that more context is needed to
decide on the formatting of a token. The formatting decision
is therefore now stored in the AttributeToken.

Kharonus commented on app/models/work_package_types/patterns/token_property_mapper.rb:
```diff
@@ -110,7 +120,12 @@ def prefixed_label(context, attribute_label)
+          formatter = if multiple
+                        ARRAY
+                      else
+                        ->(v) { v.is_a?(Symbol) ? v : v&.to_s }
```
❓ Why do we need a separate formatter for custom fields? couldn't it be `STRING_OR_NIL`? or, even better, check the type and map it, because we also have date custom fields, right?

NobodysNightmare replied:
> because we also have date custom fields, right?

Do we? That's a good point... Oh dear...

Essentially we needed the special formatter here, because we need to pass the `:attribute_not_available` through unchanged. That's why the special case for symbols.

I was hoping that we'd cover all other cases with `to_s`, but maybe I was wrong with that.

NobodysNightmare replied:
I added a test case for date custom fields. Interestingly they only support _date_, not _time_. Thus they work as intended only with `to_s`.

NobodysNightmare replied:
In other words: Until proven otherwise (ideally through a test case that we add), I'd say that the current implementation covers all cases that we need. The handling of the symbol return value is a special case specific for custom fields, which is why I wanted to locate the formatter close to that. Though we can also put it on top and name it specifically (ideas welcome), to indicate that.

Kharonus reviewed (changes requested):
I like the new approach much more, than the previous one.

But didn't we forget to remove the `stringify` method in the pattern resolver?

NobodysNightmare commented:
> But didn't we forget to remove the stringify method in the pattern resolver?

It was simplified quite a lot at least. I was once trying to remove it entirely, but it still makes sense.

We are pretty close to `token.call(context) || nil_replacement(token)`, but that would not properly consider the `:attribute_not_available` case.

If you have a better suggestion for the name, I am all ears. To be honest, the method is not a lot about stringification anymore...

Kharonus commented:
> It was simplified quite a lot at least. I was once trying to remove it entirely, but it still makes sense.

Agreed. Name can be kept for now, if someone boyscouts it eventually, it could be renamed, but I can't come up with a real smart suggestion right now.
